cluster.go

```diff
@@ -1046,13 +1046,13 @@ func (c *Cluster) isClosest(ctx context.Context, cid cid.Cid) (bool, error) {
 		return false, err
 	}
 
-	cidKey := cid.KeyString()
+	cidKey := string(kbucket.ConvertKey(cid.Hash().B58String()))
 	for _, p := range peers {
 		if p == c.id || !c.consensus.IsTrustedPeer(ctx, p) {
 			continue
 		}
 
-		if !kbucket.Closer(p, c.id, cidKey) {
+		if !kbucket.Closer(c.id, p, cidKey) {
```

Comment:
Have a look to `Closer` implementation. It already converts keys. Have you seen `SortClosestPeers()`? We just need to closest peer so maybe you can just use other available functions to sort depending on what the library exposes.

Please study what a `Cid` is, the cost of Hash().B58Stringing(). Why would you need a B58 encoded string at all? If anything you need Bytes() (and then cast to string as necessary).

Reply:
Compared to `SortClosestPeers`, one we have is more efficient. We return when we find at least one peer greater than us.

Reply:
To find whether we are closest or not, in worst case we take O(n)
Go's sort function has complexity of O(nlogn)

Reply:
Thanks for bringing that up. Now please, benchmark it and you will see that in the worst case (where you do N comparisons), your approach is slower for at least up to 1000 peers. 

On the average case (when you do O(n/2)), your function is only slightly faster) with less than 20 peers. And only twice as fast with 1000 peers.

For any reasonable sized cluster you managed to lose all your O(n) advantage in the worst case and reduce it by 80% on the average case. How much time do I have to spend telling you the way you did it is very wrong apart from more complex? I'm awfully tired of this.

And the worst is that coming up with an O(n) solution which is 10x more performant is a matter of seconds.

Reply:
A few more thoughts:

* We can use a faster function than sha256 (blake2b) is twice as fast
* The average case is actually not n/2 calls to Closer(), it's less (more the probability of finding a peer closer than youself), although always one peer will hit the worse case for each cid. This makes the current solution better in comparison to SortCloserPeers, than i assumed, but still bad because of the repeated hashing.
* IsTrustedPeer checks a syncmap in crdts. We need to benchmark what is faster, to do this 1000 times or to produce a list of trusted peers up front and loop it (with the necessary interface changes to consensus).
* The hashing operation should happen at most once for each thing (need a cache for peer id hashes).
* The whole function needs to be split and refactored

This path is perf critical because this runs regularly on potentially very large pinsets.

Reply:
> to produce a list of trusted peers up front and loop it 

I have used a map. It is faster, often twice as fast.